File: sdks/c++/basys.sdk.cc/src/aas/backend/connected/aas/submodelelement/property/ConnectedSingleProperty.cpp
/*
 * ConnectedSingleProperty.cpp
 *
 *      Author: wendel
 */

#include "ConnectedSingleProperty.h"
#include "aas/submodelelement/property/IProperty.h"

namespace basyx {
namespace aas {
namespace backend {
namespace connected {

using namespace submodelelement::property;

ConnectedSingleProperty::ConnectedSingleProperty(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedProperty(PropertyType::Single, proxy)
{}

basyx::any ConnectedSingleProperty::get() const
{
  return this->retrieveObject();
}

void ConnectedSingleProperty::set(const basyx::any & value)
{
  this->setProxyValue(PropertyPaths::VALUE, value);
}

std::string ConnectedSingleProperty::getValueType() const
{
  auto map = this->getProxy()->readElementValue("").Get<basyx::objectMap_t>();
  return map.at(PropertyPaths::VALUETYPE).Get<std::string>();
}


}
}
}
}

File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/connected/submodelelement/dataelement/property/ConnectedSingleProperty.java
package org.eclipse.basyx.submodel.metamodel.connected.submodelelement.dataelement.property;

import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.submodelelement.dataelement.property.ISingleProperty;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.dataelement.property.PropertyType;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.dataelement.property.Property;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.dataelement.property.valuetypedef.PropertyValueTypeDefHelper;
import org.eclipse.basyx.vab.exception.provider.ProviderException;
import org.eclipse.basyx.vab.modelprovider.VABElementProxy;

/**
 * Connects to a PropertySingleValued as specified by DAAS containing a simple
 * value
 * 
 * @author schnicke
 *
 */
public class ConnectedSingleProperty extends ConnectedProperty implements ISingleProperty {

	public ConnectedSingleProperty(VABElementProxy proxy) {
		super(PropertyType.Single, proxy);
	}

	@Override
	public Object get() throws Exception {
		return retrieveObject();
	}

	@Override
	public void set(Object newValue) throws ProviderException {
		getProxy().setModelPropertyValue(Property.VALUE, newValue);
	}

	@SuppressWarnings({ "unchecked" })
	@Override
	public String getValueType() {
		Object o = getProxy().getModelPropertyValue("");
		return PropertyValueTypeDefHelper.readTypeDef(((Map<String, Object>) o).get(Property.VALUETYPE)).toString();
	}

}
